Replace Extents with Size in VoxelGI, ReflectionProbe, FogVolume, Decal  and GPUParticles*3D

Maran23 opened the pull request:
I gave it a shot and revived the PR from @ator-dev.

Fixes: https://github.com/godotengine/godot/issues/54449
Supersedes: https://github.com/godotengine/godot/pull/55178
For more context see also: https://github.com/godotengine/godot/pull/44183 and https://github.com/godotengine/godot/pull/49672

Testing appreciated!

Co-authored-by: ator-dev <dominic.codedeveloper@gmail.com>

Maran23 committed:
Replace Extents with Size in VoxelGI, ReflectionProbe, FogVolume, Decal and GPUParticles*3D

- Extents are replaced by Size (Size is Extents * 2)
- The UI text displays 'Size'
- Snapping is adjusted to work with Size
- _set and _get handle extents for compatibility

Co-authored-by: ator-dev <dominic.codedeveloper@gmail.com>

Maran23 commented on editor/plugins/node_3d_editor_gizmos.cpp:
```diff
@@ -3491,8 +3491,8 @@ void ReflectionProbeGizmoPlugin::set_handle(const EditorNode3DGizmo *p_gizmo, in
 
-		extents[p_id] = d;
-		probe->set_extents(extents);
+		size[p_id] = d * 2;
```
Note: May look weird at first but is needed.
The value `d` was set before as `extents`, and since we now use `size`, which is the `extents` * 2, we need to multiply by 2 here as well.
This also follows the same pattern as used here: https://github.com/godotengine/godot/pull/44183

I just wonder if we should multiply by 2 above, when assigning `d`. (`float d = ra[p_id] * 2`)
So the snapping and `0.001` is checked against the final `size` value. 
If the answer is yes, we need to change this on other locations as well.

YuriSizov replied:
Might be worth adding a code comment if you think that line can be confusing.

Maran23 replied:
Agreed. Will wait for other opinions about the above described behaviour with `d` first.
Maybe some cleanup is needed in a follow up PR, in which case I would add a comment to all places (e.g. also the code places touched in https://github.com/godotengine/godot/pull/44183)

Maran23 replied:
Okay so this indeed does change behaviour: The snapping (holding `SHIFT`) is now done with a step of 2. Will change all locations now to restore this behaviour.

clayjohn commented:
I'm writing an early report of some issues I noticed while testing locally, I will provide more detailed comments on the code itself as well

_When dragging a handle on the ReflectionProbe the editor popup is way off. It reports the extents instead of size. The extents are labelled "P", the position is labelled "S" and the entire thing is labelled "Size X" (when dragging the "X" handle). This should just say "Size: " and report the size_
![Screenshot from 2023-01-26 12-53-26](https://user-images.githubusercontent.com/16521339/214948216-7c602145-84fc-465b-883c-9e09bf897f18.png)

_Similarly with FogVolumes "Size" reports the extents_
![Screenshot from 2023-01-26 12-51-27](https://user-images.githubusercontent.com/16521339/214948223-e75225f1-029f-4569-9184-954be7f2a4f9.png)
_Actual size for comparison_
![Screenshot from 2023-01-26 12-51-32](https://user-images.githubusercontent.com/16521339/214948221-61153a2e-dea1-489e-8205-1e4c3e20d851.png)

_FogVolume's gizmo is way off_
![Screenshot from 2023-01-26 12-51-14](https://user-images.githubusercontent.com/16521339/214948228-35fc1ac9-3afe-413e-8ce0-6e3d95b548e1.png)

I also noticed that this change introduces a shader compilation error into FogVolumes. I'll provide specific suggestions on the code momentarily.

clayjohn commented on editor/plugins/node_3d_editor_gizmos.cpp:
```diff
@@ -3460,7 +3460,7 @@ String ReflectionProbeGizmoPlugin::get_handle_name(const EditorNode3DGizmo *p_gi
 Variant ReflectionProbeGizmoPlugin::get_handle_value(const EditorNode3DGizmo *p_gizmo, int p_id, bool p_secondary) const {
 	ReflectionProbe *probe = Object::cast_to<ReflectionProbe>(p_gizmo->get_node_3d());
-	return AABB(probe->get_extents(), probe->get_origin_offset());
+	return AABB(probe->get_size() / 2, probe->get_origin_offset());
```
This isn't your fault, but the arguments are backwards here, they should be reversed as ``AABB()`` takes position then size. Also, since the label says "Size" now, this should return the size, not the extents
```suggestion
	return AABB(probe->get_origin_offset(), probe->get_size());
```

Maran23 replied:
Yeah I already figured, thanks! Should be fixed now.

clayjohn commented on editor/plugins/node_3d_editor_gizmos.cpp:
```diff
@@ -3767,19 +3767,19 @@ int VoxelGIGizmoPlugin::get_priority() const {
 Variant VoxelGIGizmoPlugin::get_handle_value(const EditorNode3DGizmo *p_gizmo, int p_id, bool p_secondary) const {
 	VoxelGI *probe = Object::cast_to<VoxelGI>(p_gizmo->get_node_3d());
-	return probe->get_extents();
+	return probe->get_size() / 2;
```
```suggestion
	return probe->get_size();
```
Same reason as above, if the label says size this should return size not extents

clayjohn commented on editor/plugins/node_3d_editor_gizmos.cpp:
```diff
@@ -5902,11 +5902,11 @@ int FogVolumeGizmoPlugin::get_priority() const {
 
 Variant FogVolumeGizmoPlugin::get_handle_value(const EditorNode3DGizmo *p_gizmo, int p_id, bool p_secondary) const {
-	return Vector3(p_gizmo->get_node_3d()->call("get_extents"));
+	return Vector3(p_gizmo->get_node_3d()->call("get_size")) / 2;
```
Same as above
```suggestion
	return Vector3(p_gizmo->get_node_3d()->call("get_size"));
```

clayjohn commented on editor/plugins/node_3d_editor_gizmos.cpp:
```diff
@@ -5968,8 +5968,8 @@ void FogVolumeGizmoPlugin::redraw(EditorNode3DGizmo *p_gizmo) {
 		AABB aabb;
-		aabb.position = -cs->call("get_extents").operator Vector3();
-		aabb.size = aabb.position * -2;
+		aabb.size = -cs->call("get_size").operator Vector3();
+		aabb.position = aabb.size / 2;
```
This seems mixed up. Its likely responsible for the broken gizmo in my screenshot
```suggestion
		aabb.size = cs->call("get_size").operator Vector3();
		aabb.position = -aabb.size / 2;
```

Maran23 replied:
Fixed

clayjohn commented on scene/3d/reflection_probe.cpp:
```diff
@@ -85,38 +85,38 @@ float ReflectionProbe::get_mesh_lod_threshold() const {
 
-		if (extents[i] - 0.01 < ABS(origin_offset[i])) {
-			origin_offset[i] = SIGN(origin_offset[i]) * (extents[i] - 0.01);
+		if (size[i] - 0.01 < ABS(origin_offset[i])) {
+			origin_offset[i] = SIGN(origin_offset[i]) * (size[i] - 0.01);
```
I think this logic has changed. My reading of this is that it is clamping the origin offset to be within the extents. However, with this change origin offset will only be clamped within a range of size from the origin i.e. within 2*size. So this logic needs to be updated to ensure that the behaviour stays the same

_edit: Ah, you made the change already below_
```
if (size[i] / 2 - 0.01 < ABS(origin_offset[i])) {
	origin_offset[i] = SIGN(origin_offset[i]) * size[i] / 2 - 0.01;
```

clayjohn commented on scene/3d/voxel_gi.cpp:
```diff
@@ -273,14 +273,14 @@ VoxelGI::Subdiv VoxelGI::get_subdiv() const {
-	extents = Vector3(MAX(1.0, p_extents.x), MAX(1.0, p_extents.y), MAX(1.0, p_extents.z));
+void VoxelGI::set_size(const Vector3 &p_size) {
+	// Prevent very small size dimensions as these breaks baking if other size dimensions are set very high.
+	size = Vector3(MAX(1.0, p_size.x), MAX(1.0, p_size.y), MAX(1.0, p_size.z));
```
I'm skeptical about this. This effectively decreases the minimum size by half

Maran23 replied:
Yeah that was my thought as well. So we should probably use `0.5` here, right?

clayjohn replied:
Okay, should be fine. If I recall correctly the minimum size was chosen rather arbitrarily

clayjohn commented on scene/resources/fog_material.cpp:
```diff
@@ -159,7 +159,7 @@ uniform sampler3D density_texture: hint_default_white;
     DENSITY = density * clamp(exp2(-height_falloff * (WORLD_POSITION.y - OBJECT_POSITION.y)), 0.0, 1.0);
     DENSITY *= texture(density_texture, UVW).r;
-    DENSITY *= pow(clamp(-SDF / min(min(EXTENTS.x, EXTENTS.y), EXTENTS.z), 0.0, 1.0), edge_fade);
+    DENSITY *= pow(clamp(-SDF / min(min(SIZE.x, SIZE.y), SIZE.z) / 2, 0.0, 1.0), edge_fade);
```
```suggestion
    DENSITY *= pow(clamp(-2.0 * SDF / min(min(SIZE.x, SIZE.y), SIZE.z), 0.0, 1.0), edge_fade);
```
You can't divide a float by an integer in GLSL, also a division by a division is a multiplication, so switching this to use multiplication by 2.0 should solve the bug and make it slightly faster

Maran23 replied:
Sorry, I don't quite get that.
How is `-SDF / X / 2.0` the same as `-2.0 * SDF / X`?

clayjohn replied:
It works if you imagine parentheses around ``min(min(SIZE.x, SIZE.y), SIZE.z) / 2,`` like I somehow did haha!

You are right, it should be ``-0.5 * SDF / min(min(SIZE.x, SIZE.y), SIZE.z)``

Maran23 replied:
Ahh, yes, makes sense. Will change it. :)

clayjohn commented on servers/rendering/renderer_rd/environment/fog.cpp:
```diff
@@ -516,7 +516,7 @@ void Fog::volumetric_fog_update(const VolumetricFogSettings &p_settings, const P
 
 		Vector2 frustum_near_size = p_cam_projection.get_viewport_half_extents();
-		Vector2 frustum_far_size = p_cam_projection.get_far_plane_half_extents();
+		Vector2 frustum_far_size = p_cam_projection.get_viewport_half_extents();
```
I'm not sure why this change was made it should likely be reverted

clayjohn commented on servers/rendering/renderer_rd/environment/fog.cpp:
```diff
@@ -683,9 +683,9 @@ void Fog::volumetric_fog_update(const VolumetricFogSettings &p_settings, const P
-			push_constant.extents[0] = extents.x;
-			push_constant.extents[1] = extents.y;
-			push_constant.extents[2] = extents.z;
+			push_constant.size[0] = extents.x;
+			push_constant.size[1] = extents.y;
+			push_constant.size[2] = extents.z;
```
```suggestion
			push_constant.size[0] = extents.x * 2.0;
			push_constant.size[1] = extents.y * 2.0;
			push_constant.size[2] = extents.z * 2.0;
```

Maran23 commented:
> I'm writing an early report of some issues I noticed while testing locally, I will provide more detailed comments on the code itself as well

Thanks! I will take a look at it. I'm not super familiar with the 3D part of the engine, much more on the 2D part so this is super helpful!

clayjohn commented on servers/rendering/renderer_rd/environment/fog.cpp:
```diff
@@ -966,8 +966,8 @@ void Fog::volumetric_fog_update(const VolumetricFogSettings &p_settings, const P
 
-	Vector2 frustum_near_size = p_cam_projection.get_viewport_half_extents();
-	Vector2 frustum_far_size = p_cam_projection.get_far_plane_half_extents();
+	Vector2 frustum_near_size = p_cam_projection.get_viewport_half_extents() * 2.0;
+	Vector2 frustum_far_size = p_cam_projection.get_far_plane_half_extents() * 2.0;
```
This seems odd, you are double the near and far size but not making an equivalent change in the shader, so this is likely to change behaviour. Likely needs to be reverted

Maran23 replied:
Yeah, I think I overlooked it. Thanks!

clayjohn commented on doc/classes/RenderingServer.xml:
```diff
@@ -1239,7 +1231,15 @@
 			<param index="1" name="shape" type="int" enum="RenderingServer.FogVolumeShape" />
 			<description>
-				Sets the shape of the fog volume to either [constant RenderingServer.FOG_VOLUME_SHAPE_ELLIPSOID], [constant RenderingServer.FOG_VOLUME_SHAPE_CONE], [constant RenderingServer.FOG_VOLUME_SHAPE_CYLINDER], [constant RenderingServer.FOG_VOLUME_SHAPE_BOX] or [constant RenderingServer.FOG_VOLUME_SHAPE_WORLD].
+				Sets the shape of the fog volume to either [constant RenderingServer.FOG_VOLUME_SHAPE_ELLIPSOID], [constant RenderingServer.FOG_VOLUME_SHAPE_CONE], [constant RenderingServer.FOG_VOLUME_SHAPE_CYLINDER] or [constant RenderingServer.FOG_VOLUME_SHAPE_BOX].
```
This looks like an unintentional change

Maran23 replied:
Yes, thanks for catching it!

clayjohn commented on doc/classes/ReflectionProbe.xml:
```diff
@@ -43,7 +39,7 @@
 		<member name="max_distance" type="float" setter="set_max_distance" getter="get_max_distance" default="0.0">
 			The maximum distance away from the [ReflectionProbe] an object can be before it is culled. Decrease this to improve performance, especially when using the [constant UPDATE_ALWAYS] [member update_mode].
-			[b]Note:[/b] The maximum reflection distance is always at least equal to the [member extents]. This means that decreasing [member max_distance] will not always cull objects from reflections, especially if the reflection probe's [member extents] are already large.
+			[b]Note:[/b] The maximum reflection distance is always at least equal to the [member size]. This means that decreasing [member max_distance] will not always cull objects from reflections, especially if the reflection probe's [member size] are already large.
```
```suggestion
			[b]Note:[/b] The maximum reflection distance is always at least equal to the probe's extents. This means that decreasing [member max_distance] will not always cull objects from reflections, especially if the reflection probe's [member size] is already large.
```

clayjohn reviewed (commented):
I have left some comments and suggestions.

We need to also consider updating Decals, GPUParticlesCollisionBox3D, GPUParticlesCollisionSDF3D, GPUParticlesCollisionHeightField3D, GPUParticlesAttractorBox3D, and GPUParticlesAttractorVectorField3D while we are at it.

Maran23 commented:
Update:
1. I fixed a lot of things I found. 
2. I added `Decal` now as well, since the logic very close to the rest.
3. I wonder if it makes sense to change the `GPUParticles...` nodes in another PR, so this PR does not grow much further?
I have no strong opinion, I can also add them here. :)

Calinou commented:
> I wonder if it makes sense to change the GPUParticles* nodes in another PR, so this PR does not grow much further?
I have no strong opinion, I can also add them here. :)

Yes, it makes sense to do this in the same PR.

Maran23 commented:
> Yes, it makes sense to do this in the same PR.

Changed all the `GPUParticles*3D` as well.

Maran23 commented on servers/rendering/renderer_rd/renderer_scene_render_rd.h:
```diff
@@ -204,9 +204,9 @@ class RendererSceneRenderRD : public RendererSceneRender {
+	virtual void setup_added_reflection_probe(const Transform3D &p_transform, const Vector3 &p_half_size){};
 	virtual void setup_added_light(const RS::LightType p_type, const Transform3D &p_transform, float p_radius, float p_spot_aperture){};
-	virtual void setup_added_decal(const Transform3D &p_transform, const Vector3 &p_half_extents){};
+	virtual void setup_added_decal(const Transform3D &p_transform, const Vector3 &p_half_size){};
```
Note that `p_half_extents` was actually named wrong before. The `extents` (unchanged) were always passed in, which is half the size.

akien-mga commented on scene/resources/fog_material.cpp:
```diff
@@ -159,7 +159,7 @@ uniform sampler3D density_texture: hint_default_white;
     DENSITY = density * clamp(exp2(-height_falloff * (WORLD_POSITION.y - OBJECT_POSITION.y)), 0.0, 1.0);
     DENSITY *= texture(density_texture, UVW).r;
-    DENSITY *= pow(clamp(-SDF / min(min(EXTENTS.x, EXTENTS.y), EXTENTS.z), 0.0, 1.0), edge_fade);
+    DENSITY *= pow(clamp(-0.5 * SDF / min(min(SIZE.x, SIZE.y), SIZE.z), 0.0, 1.0), edge_fade);
```
Shouldn't this be `2.0` if `SIZE` is on the denominator?

Maran23 replied:
No that won't work. Actually @clayjohn had the same idea first:
![image](https://user-images.githubusercontent.com/66004280/215819581-bc229eb9-0e7d-428e-a56f-834b419eeefa.png)

akien-mga replied:
Well, the original code was:
```
-SDF / EXTENTS
```
Now we have `EXTENTS = 0.5 * SIZE`, so:
```
-SDF / (0.5 * SIZE)
```

Multiplying both sides with 2, it gives:
```
-2.0 * SDF / SIZE
```

Maran23 replied:
Ahh good catch, I see the problem. Yes, this is right. Damn braces. 😄

clayjohn commented on scene/resources/fog_material.cpp:
```diff
@@ -159,7 +159,7 @@ uniform sampler3D density_texture: hint_default_white;
     DENSITY = density * clamp(exp2(-height_falloff * (WORLD_POSITION.y - OBJECT_POSITION.y)), 0.0, 1.0);
     DENSITY *= texture(density_texture, UVW).r;
-    DENSITY *= pow(clamp(-SDF / min(min(EXTENTS.x, EXTENTS.y), EXTENTS.z), 0.0, 1.0), edge_fade);
+    DENSITY *= pow(clamp(-2 * SDF / min(min(SIZE.x, SIZE.y), SIZE.z), 0.0, 1.0), edge_fade);
```
```suggestion
    DENSITY *= pow(clamp(-2.0 * SDF / min(min(SIZE.x, SIZE.y), SIZE.z), 0.0, 1.0), edge_fade);
```
Needs to be a float or the shader will fail to compile

Maran23 replied:
> Do you have time in the next 14 hours or so to add this? If not I will happily do it

Yes, will working on it right now. :)

Maran23 replied:
Damn, thanks!

clayjohn commented:
@Maran23 We discussed this PR in our production meeting this morning. The consensus was that we would like to merge ithis ASAP as we don't want this work to go to waste and it is kind of now or never. However, we also want to minimize compatibility breakage (after all, we did promise to minimize compatibility breakage after Beta 1). 

To minimize compatibility breakage, we need to add glue code to the ``_set/_get`` functions that can read when extents is used and automatically set size accordingly. I.e.:
```
#ifndef DISABLE_DEPRECATED
bool Decal::_set(const StringName &p_name, const Variant &p_value) {
	if (p_name == "extents") {
		set_size(p_value * 2.0);
		return true;
	}
	return false;
}

bool Decal::_get(const StringName &p_name, Variant &r_ret) const {
	if (p_name == "extents") {
		r_ret = get_size() / 2.0;
		return true;
	}
	return false;
}
#endif
```

Do you have time in the next 14 hours or so to add this? If not I will happily do it

Maran23 commented:
> Do you have time in the next 14 hours or so to add this? If not I will happily do it

I will work on it right now. :)

clayjohn commented:
Here is a test project built in Beta 16 that I will be using to check if extents are properly converted from save files 
[extents-to-size.zip](https://github.com/godotengine/godot/files/10550008/extents-to-size.zip)

Maran23 commented:
> Here is a test project built in Beta 16 that I will be using to check if extents are properly converted from save files [extents-to-size.zip](https://github.com/godotengine/godot/files/10550008/extents-to-size.zip)
> 
> Also tested GDScript code that sets `.extents` directly

Thanks for the project.
I opened it with Godot 16 and a copy with that branch.
Looks good to me:
![image](https://user-images.githubusercontent.com/66004280/215880079-f336082e-d98c-4a79-b1ea-afcb5624ac6f.png)

clayjohn reviewed (approved):
Seems good to me after reviewing and testing locally. This should minimize compat breakage. The only problematic situation will be if users use ``set_extents()`` from script.

akien-mga commented:
Thanks!
